### ISSUE TITLE
base-hw scheduler: modernize internal algorithm

### ISSUE THREAD

**Issue by m-stein** [genodelabs/genode#4700]
As already announced in https://github.com/genodelabs/genode/issues/4151#issuecomment-1333605604, I'd like to propose a modernized implementation for the base-hw scheduler. This is my idea, so far, after re-thinking it several times and discussing it with @chelmuth:

```
Prioritized, Curr Round
Round ID: 6
 |
 |          Prio Bands Array
 |          +-----+-----+-----+----+-----+---+---+
 +--------->| 127 | 126 | ... | 98 | ... | 1 | 0 |
 |          +-----+-----+-----+----+-----+---+---+
 |                   |          |          |
 |         +---------+          +-----+    +---------------------+
 |         |                          |                          |
 |         V                          V                          V
 |     +---------+   +--+   +--+   +--------+   +--+   +--+   +-------+   +--+
 +---->| 10/20   |<->|  |<->|  |<->| 2/10   |<->|  |<->|  |<->| 9/30  |<->|  |
       | R6 P126 |   |  |   |  |   | R6 P98 |   |  |   |  |   | R6 P1 |   |  |
       +---------+   +--+   +--+   +--------+   +--+   +--+   +-------+   +--+

Prioritized, Next Round
Round ID: 7
 |
 |          Prio Bands Array
 |          +-----+-----+-----
 +--------->| 127 | 126 | ...
 |          +-----+-----+-----
 |             |
 |         +---+
 |         |
 |         V
 |     +---------+   +--+
 +---->| 40/40   |<->|  |<-> ...
       | R7 P127 |   |  |
       +---------+   +--+

Unprioritized
Remaining time of head: 12
 |
 |     +--+   +--+   +--+   +--+   +--+
 +---->|  |<->|  |<->|  |<->|  |<->|  |
       +--+   +--+   +--+   +--+   +--+
```

The scheduler holds two identical data structures for the prioritized scheduling, named "Current Round" and "Next Round" and one for the unprioritized scheduling, named "Unprioritized". Unprioritized is a list plus the remaining time slice of the head of that list (no more ``_fill`` members in the scheduling contexts). It's a simple Round-Robin.

Current Round contains a list of all scheduling contexts with priority that are ready and have quota left for the current round (superperiod). This list is sorted by priority from the highest to the lowest. There is a fixed number of priority bands (considerably higher than before). In addition to the list, there is an array of pointers into that list and the array indices are the priority bands. Each array item references the first list item with this priority. The array is for fast insertion of a scheduling context with a certain priority. The list is for fast determination of the scheduling choice.

When the Current Round list is not empty, the head of that list is the current scheduling choice. When the Current Round list is empty but the Unprioritized list is not, the head of that latter list is the scheduling choice. Whenever the two lists are empty, the idle context is scheduled.

Each prioritized context holds the Round ID of its remaining quota. When the remaining quota of a scheduling context in Current Round gets depleted, its Round ID is incremented and the context is moved to Next Round. Next Round is organized the same way as Current Round. The only difference is to which round the remaining quota and round ID of the contexts corresponds.

When a context becomes unready, the scheduler simply removes it from all its data structures (fast because of double lists). When a context becomes ready, the scheduler compares the context's Round ID to the Current Round ID. If it is lower, the context's remaining quota is reset and the context is added to Current Round. If it is the same, the context is added to the Current Round. If it is one higher, the context is added to Next Round.

Characteristics of this approach would be the following:

* There is no "unready contexts" or "present contexts" list which complicated the logic in the old scheduler.
* No operation on the scheduler has an overhead that depends on the number of used priority bands.
* Using more priority bands only raises the static memory requirement of the kernel (in the above depicted example, assuming 4 CPU cores, the kernel would need 8Kb for the priority-band arrays).
* No separation of context operations (set ready, set unready, set quota, yield, etc.) and the scheduler update, i.e., no buffered and potentially outdated scheduling choices. At the end of each context operation, the internal state of the scheduler efficiently yields the correct scheduling choice. Simplifies using the scheduler and makes it less error-prone.
* No dynamic memory allocation and the size of a scheduling context actually decreases compared to the old scheduler.

**Comment by m-stein** [comment]
I forgot to mention:

When a Round ends, the remaining contexts in the Current Round list are reset and moved to Next Round. Then Next Round becomes Current Round and the former Current Round becomes Next Round with the round ID increased by 2.

**Comment by m-stein** [comment]
This can probably be simplified further by connecting the Unprioritized list directly to the tail of the Current Round list. This would also align well with a requirement that was communicated recently to me that prioritized contexts become the head of the Unprioritized list in the moment when their quota gets depleted.
```
Prioritized, Curr Round
Round ID: 6
Remaining time of unprioritized head: 12
 |
 |          Prio Bands Array
 |          +-----+-----+-----+----+-----+---+---+                        +---------------+
 +--------->| 127 | 126 | ... | 98 | ... | 1 | 0 |                        | Unprioritized |
 |          +-----+-----+-----+----+-----+---+---+                        +---------------+
 |                   |          |          |                                      |
 |         +---------+          +-----+    +---------------------+                |
 |         |                          |                          |                |
 |         V                          V                          V                V
 |     +---------+   +--+   +--+   +--------+   +--+   +--+   +-------+   +--+   +--+   +--+   +--+   +--+   +--+
 +---->| 10/20   |<->|  |<->|  |<->| 2/10   |<->|  |<->|  |<->| 9/30  |<->|  |<->|  |<->|  |<->|  |<->|  |<->|  |
       | R6 P126 |   |  |   |  |   | R6 P98 |   |  |   |  |   | R6 P1 |   |  |   +--+   +--+   +--+   +--+   +--+
       +---------+   +--+   +--+   +--------+   +--+   +--+   +-------+   +--+
```
However, then it's probably the best to manage the Next Round list in parallel with a second list item in each context. This would also eliminate the effort of moving all remaining Current Round contexts at the end of the round and thereby further mitigate overhead peaks.

**Comment by m-stein** [comment]
I think it's safe to assume that this issue was superseded by #5117.
